doc/cmis-module-enhancement/cmis-module-enhancement.md

```diff
@@ -0,0 +1,224 @@
+It will ensure that the module initialization doesn't start before the high-speed signal is transmitted by ASIC to a module.
+
+The notification shall be expected and consumed only on platforms supporting it. 
+Hence, The platform capabilities for supporting of this feature should be queried on Port OA init.
+
+On platforms not supporting this functionality the “host_tx_ready” flag shall be set in STATE DB upon return of Port Admin status UP with SUCCESS return code (backward-compatible behavior)
+
+<p align=left>
+<img src="img/Host-Tx-Ready-Flow.png" alt="Figure 3. Host Tx Signal Enable Flow">
+</p>
+
+For supporting platforms, admin status UP is not sufficient for configuring the module.
+Ports OA will set host_tx_ready in state DB only after it knows a module was plugged, and received notification from SAI of the high-speed signal.
+When the high-speed signal is sent, a notificatoin from SAI named "SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY" will arrive to swss. 
+PortsOrch will update State DB with host_tx_ready, after the notification has arrived.
+
+Cmis Manager task behavior will stay the same as today - start configuring the module only after admin state is UP, and host_tx_ready=true.
+
+### 6.3. Implementation Flow
+
+#### 6.3.1. Initialization Flow
+
+Proposed enhancement will be supported on some vendor platforms, so there is a need to learn whether the specific platform supports them on not.
+To do that, Ports OA shall use the query capability for the following SAI attributes:
+
+1.	 SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE – to know whether the platform supports the control of enabling/disabling the high-speed signal from ASIC to module from SONIC.
+2.	SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY – to know whether the platform supports the asynchronous notification from SDK/SAI to SWSS about start/stop of transmission of high-speed signal data from ASIC to module.
+
+Ports OA shall use the status of support of these capabilities for its flow.
+For example, the notification consumer for a new SAI_SWITCH_ATTR_PORT_HOST_TX_READY_NOTIFY notification will be initialized only if the platform supports the asynchronous notification from SAI/SDK to SWSS.
+
+#### 6.3.2. Enabling/disabling “host_tx_signal”
+As described earlier, for platforms supporting this capability, high-speed signal shall be explicitly enabled/disabled by Ports OA on getting the Insertion/Removal event for a specific module.
+
+When a module is plugged in/out the PMON adds/deletes a per-port entry in the TRANSCIEVER_TABLE in STATE DB.
+This event shall be used by Ports OA as a trigger for enabling or disabling the host_tx_signal (using setting true/false to SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE on the Port Object).
+
+To detect this insertion/removal event the Ports Orch Agent shall register for changes in the TRANSCIEVER_TABLE in STATE DB and will do the following:
```

Comment:
@noaOrMlnx AFAIK, STATE_DB TRANSCEIVER_TABLE is not preserved across warm-reboot. If so, my concern is during warm-reboot if the TRANSCEIVER_TABLE gets deleted, then SAI_PORT_ATTR_HOST_TX_SIGNAL_ENABLE could become FALSE after warm-reboot. Could you please check this.?

Reply:
@prgeor 
Hi, 
When we execute warm-reboot, the whole table is deleted. Since we are performing a reboot, it will happen after OA will stop, so no notification is received. 
When OA is up again after reboot, the table is clean, so notification will be received only upon new info inserted to the table. 
So for me it looks like we should not worry about it.